rust/connlib/tunnel/src/sockets.rs

```diff
@@ -292,44 +293,55 @@ impl Socket {
     }
 
     fn poll_flush(&mut self, cx: &mut Context<'_>) -> Poll<io::Result<()>> {
-        loop {
-            match self.socket.try_io(Interest::WRITABLE, || {
-                self.state
-                    .send((&self.socket).into(), &self.buffered_transmits)
-            }) {
-                Ok(0) => break,
-                Err(e) if e.kind() == io::ErrorKind::WouldBlock => break,
-                Err(e) => return Poll::Ready(Err(e)),
-
-                Ok(num_sent) => {
-                    self.buffered_transmits.drain(..num_sent);
+        while let Some(transmit) = self.buffered_transmits.front() {
+            ready!(self.socket.poll_send_ready(cx))?;
 
-                    // I am not sure if we'd ever send less than what is in `buffered_transmits`.
-                    // loop once more to be sure we `break` on either an empty buffer or on `WouldBlock`.
+            match self.try_send(transmit) {
+                Ok(()) => {
+                    self.buffered_transmits.pop_front();
                 }
-            };
+                Err(e) if e.kind() == io::ErrorKind::WouldBlock => continue, // False positive wake-up: Loop to `poll_send_ready` and return `Pending`.
+                Err(e) => return Poll::Ready(Err(e)),
+            }
         }
 
-        // Ensure we are ready to send more data.
-        ready!(self.socket.poll_send_ready(cx)?);
-
-        assert!(
-            self.buffered_transmits.is_empty(),
-            "buffer must be empty if we are ready to send more data"
-        );
+        assert!(self.buffered_transmits.is_empty());
 
         Poll::Ready(Ok(()))
     }
 
-    fn send(&mut self, transmit: quinn_udp::Transmit) {
-        tracing::trace!(target: "wire::net::send", src = ?transmit.src_ip, dst = %transmit.destination, num_bytes = %transmit.contents.len());
-
-        self.buffered_transmits.push(transmit);
+    fn send(&mut self, transmit: snownet::Transmit) -> io::Result<()> {
+        tracing::trace!(target: "wire::net::send", src = ?transmit.src, dst = %transmit.dst, num_bytes = %transmit.payload.len());
 
         debug_assert!(
             self.buffered_transmits.len() < 10_000,
             "We are not flushing the packets for some reason"
         );
+
+        match self.try_send(&transmit) {
+            Ok(()) => Ok(()),
+            Err(e) if e.kind() == io::ErrorKind::WouldBlock => {
+                tracing::trace!("Buffering packet because socket is busy");
+
+                self.buffered_transmits.push_back(transmit.into_owned());
```

Comment:
So the reason `self.try_send` is in two places is, we can avoid the `into_owned` and be zero-copy if the socket happens to be ready for sending?

Reply:
Yeah that is the idea.

Reply:
This buffer could grow a lot, should we bound this buffer? The send buffer in linux by default ~200kB so I'd just say let's do it x10 that?

Reply:
> This buffer could grow a lot, should we bound this buffer? The send buffer in linux by default ~200kB so I'd just say let's do it x10 that?

I don't think it can. We suspend on flushing all items in here before reading new ones so the only ones that can accummulate are STUN messages.

Reply:
There is also that debug_assert for 10,000 items up there